solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudSolrClient.java

```diff
@@ -1228,9 +1199,15 @@ protected DocCollection getDocCollection(String collection, Integer expectedVers
       if (col != null) {
         if (expectedVersion <= col.getZNodeVersion() && !cacheEntry.shouldRetry()) return col;
       }
+      // Moving the lookup inside the lock makes the lazy/non-lazy check irrelevant
+      ClusterState.CollectionRef ref = getCollectionRef(collection);
+      if (ref == null) {
+        // no such collection exists
+        return null;
+      }
       // We are going to fetch a new version
       // we MUST try to get a new version
-      DocCollection fetchedCol = ref.get(); // this is a call to ZK
+      DocCollection fetchedCol = ref.get(); // this is a call to ZK if CSP instance is a Zk CSP
```

Comment:
The ZK-ness isn't the point; the point is that a remote call is going to be made (expensive)